### DISCUSSION TITLE
Ability to have custom relationship/collection queries

### DISCUSSION THREAD

**Discussion by rat-matheson** [mikro-orm/mikro-orm#3117] [Ideas]
**Is your feature request related to a problem? Please describe.**
When implementing a logical delete on one object, I'd like the ability to have the relationship not include the entire key from the parent object.

Example

```typescript
@Entity({
    tableName:'user'
})
export class UserEntity implements ILogicalDelete {
    @PrimaryKey({
        length:30
    })
    public id:string;

    @PrimaryKey({
        fieldName:"aud_deleted_date",
        type:TransformType.newTransform(...), // custom transformation which just interprets the DB value to appropriately update the entity
        hidden:true,
        default:0,
        comment:"If the data is not deleted, the value is 0.  Otherwise, it is the date the data was deleted" // A date is used over a boolean because sometimes things are logically deleted multiple times.  Ie, could add the user, delete, then add the same user again
    })
    public deletedDate:Date | 0 = 0;

    [PrimaryKeyType]: [string, number];

    @OneToMany({ 
        entity: () => UserFlagRel, 
        mappedBy:e => e.userId,
        cascade: [Cascade.REMOVE],
        orphanRemoval: true})
    public flags?:Collection<UserFlagRel> = new Collection<UserFlagRel>(this);
}


@Entity({
    tableName:"user_flags"
})
export class UserFlagRel {
    
    @ManyToOne({
        primary:true,
        mapToPk: true,
        entity: () => UserEntity,
    })
    public userId:string;  //ISSUE: this needs to be [string, number] to account for the entire primary key but in my actual data model, I only need the user.id part of the user.

    @PrimaryKey({
        length:64
    })
    public flag:string;

    [PrimaryKeyType]: [string, string];
}
```

**Describe the solution you'd like**
I would like a way to map the key to another value

```typescript
@Entity({
    tableName:'user'
})
export class UserEntity implements ILogicalDelete {
    @PrimaryKey({
        length:30
    })
    public id:string;

    @PrimaryKey({
        fieldName:"aud_deleted_date",
        type:TransformType.newTransform(...),
        hidden:true,
        default:0,
        comment:"If the data is not deleted, the value is 0.  Otherwise, it is the date the data was deleted" 
    })
    public deletedDate:Date | 0 = 0;

    [PrimaryKeyType]: [string, number];

    @OneToMany({ 
        entity: () => UserFlagRel, 
        mappedBy:e => ({
            id:e.userId,
            deletedDate:0
        }),
        cascade: [Cascade.REMOVE],
        orphanRemoval: true})
    public flags?:Collection<UserFlagRel> = new Collection<UserFlagRel>(this);
}


@Entity({
    tableName:"user_flags"
})
export class UserFlagRel {
    
    @ManyToOne({
        primary:true,
        mapToPk: u => u.id,
        entity: () => UserEntity,
    })
    public userId:string; 

    @PrimaryKey({
        length:64
    })
    public flag:string;

    [PrimaryKeyType]: [string, string];
}
```

**Describe alternatives you've considered**
For the specific case of logical delete, I could have just the user.id as the primary key instead of a composite key.  However, composite key makes way more sense.  It forces the primary key to include this essential logical deletion aspect, and also organizes the data on disk to match as well.

I don't think it makes sense to add a logical deletion to the user_flags table.  It adds unnecessary complexity.  I want the idea of the logical deletion to be completely encapsulated by the user entity/table.  Anything consuming that  (in my case) doesn't care about the logical deletion based on my query patterns.

**Additional context**

Here is how the queries are being generated currently since I have not yet designed the UserFlagRel to account for the logical deletion in the UserEntity (and hopefully won't have to).

```sql
 select `u0`.* from `user_flags` as `u0` where (`u0`.`user_id_id`, `u0`.`user_id_aud_deleted_date`) = 'CkGjqJZPURQ9P5GL0atoL' order by (`u0`.`user_id_id`, `u0`.`user_id_aud_deleted_date`) asc, `u0`.`flag` asc limit 1001
```

**Comment by rat-matheson** [comment]
Related to the above, perhaps I am looking for a way to populate a collection with a custom query?  That seems like a more general statement and solution to the problem I'm encountering.  As a related example, suppose of I have a table of products with columns "company, model, year, inProduction (yes/no), ...".  I might have another object, say Store and for this object, I want to have a collection of just discontinued products still available at the store.  I also need a relationship table linking the two together.

```typescript
@Entity(...)
class Store {
    ...
    @ManyToMany({
        type:() => Product,
        query:r => ({
                storeId:r.id,
                productId:() => ({
                    inProduction:false
                    })
        })
    })
    public discontinuedItems:Collection<Product> = new Collection<Product>(this);
}
```

**Comment by B4nan** [comment]
Moving to discussions before we settle on what exactly you are asking here, especially because of the OP - I am not interested in some magic over defining PKs inside relations. Not exactly sure what you refer to as "logical deletion", do you mean soft delete? There are easier ways to achieve it than making such a weird schema, [like using `onFlush` subscriber to convert delete queries to update ones](https://twitter.com/MikroOrm/status/1365019173479448582). PK is supposed to be PK, you should not change it because the entity state changed (= was soft delete), never heard of anyone doing this, is it coming from some tutorial or is it your own idea? How would relations work, you would expect every FK to be defined with `on update cascade`? The ORM holds entity state based on the PK value, so just by doing this you would break this invariant.

> Related to the above, perhaps I am looking for a way to populate a collection with a custom query?

If you want additional queries for any relation, just add them when you populate the relation, you can already add conditions via `FindOptions.populateWhere`. We could allow doing this in the entity definition, but such collection should be readonly as I am not interested in parsing your custom queries and doing runtime validation for things you could add to such collection. Alternative to this is `Collection.matching` that allows partial loading of collections (those will be also marked as readonly).

Moveover you can hydrate a collection via `Collection.hydrate` with anything. The data can even come from a raw SQL, just map them to entities via `em.map()` and pass them to the `hydrate` method.

**Comment by rat-matheson** [comment]
Thanks so much for considering my post and responding.  A discussion does seem more appropriate as it seems like it will be necessary to tease out the use case.

With regards to the logical delete, fair points above.  The part that stands out as being weirdest to me is the deletedDate variable having two purposes (which is an anti-pattern).  I will explain the reasoning for that later.   Also, because you are so familiar with mikro orm, there may be some obvious features to you that make implementing logical deletes more elegant.   I used to work as an engineer at Amazon and we implemented it this way (or similar) many times.  That is also where I picked up the term logical delete as opposed to soft delete.  I believe they are equivalent.

With regards to single table logical deletes, the storage requirement for primary key is something to uniquely identify the object (eg email address or id) and something to determine it's deleted status.  Often times, people will use an isDeleted flag.  This doesn't work for scenarios where the identifier for the object can be re-inserted in the future because there will be a primary key conflict.  It also doesn't work for scenarios where you want some level of single table auditing for deletes.  This second audit scenario happens sometimes with larger datasets and high call volume where something just gets logically deleted and then a job will remove the deleted records later storing them in a more appropriate auditing system.  

I've ended up using the deletedDate as my logically deleted flag.  I agree that it is weird in that it is a single variable containing two jobs.  This is generally an anti-pattern but it's use seems appropriate here, especially if I can completely encapsulate the variable in the dao layer.  Using 0 instead of null is something I got from some blog post that I can't find at the moment but it is very similar to what is described [here on stack overflow](https://stackoverflow.com/questions/68323/what-is-the-best-way-to-implement-soft-deletion).

> PK is supposed to be PK, you should not change it because the entity state changed (= was soft delete)

This is a good thing to consider.  The flag has to be part of the primary key because an item might be deleted at one point and then re-added again later.  Eg, imagine the primary key is [email, deletedDate].  That allows for the object to be inserted at one point, deleted, and then inserted again.  However if the pk was just email, there would be a constraint failure (or the old data would have to be actually removed from the table rather than just logically removed).

I hope that by going through those use cases, this implementation for logical delete seems a little more reasonable at least.  

Appreciate the [link you shared](https://twitter.com/MikroOrm/status/1365019173479448582) for the on flush.  Very cool feature.  In that link, it does not account for the use case where a deletedDate is used.  If we were to use that instead of the deleted flag, we end up with the issues that prompted this discussion.  Either we have a primary key conflict in the future for the same id being re-inserted, or the deletedDate is added to the primary key and now the relationships have to be away of the deletedDate status even though I never will pull anything with a non-zero deletedDate into my model.

What about this?  Would it make sense to have the primary key in the DB different from the one defined in mikro-orm?  I think that *could* solve the issue.  I just have to make sure that when I query the UserEntity, I also set the where clause to use {deletedDate:0} which I do anyway.  And then I can use the onFlush to update the deletedDate propery like you suggested in your link.

**Comment by rat-matheson** [comment]
Just to follow up, I was thinking about why the implementation uses "0" instead of null.  I suspect it is because in javascript, many things can equal null.  This can introduce unexpected bugs like the one I'm working on right now where a signature is getting screwed up because it is using null rather than 0.
